torch/csrc/generic/Storage.cpp

```diff
@@ -31,14 +34,42 @@ static PyObject * THPStorage_(pynew)(PyTypeObject *type, PyObject *args, PyObjec
   THPStoragePtr self((THPStorage *)type->tp_alloc(type, 0));
   THPUtils_assert(self, "failed to allocate a " THPStorageStr " object");
   c10::Allocator* allocator = nullptr;
+#ifdef USE_CUDA
+  at::cuda::OptionalCUDAGuard device_guard;
+#endif
 
   // Internally we allow constructing with a keywoard only argument cdata
   if (kwargs != nullptr) {
     PyObject *allocator_ptr = PyDict_GetItemString(kwargs, "allocator");
+    PyObject *device_ptr = PyDict_GetItemString(kwargs, "device");
+
+    TORCH_CHECK(!allocator_ptr || !device_ptr,
+        THPStorageStr, "(): cannot provide both 'allocator' and 'device' kwargs");
+
     if (allocator_ptr) {
       THPUtils_assert(THPUtils_checkLong(allocator_ptr), "invalid allocator");
       allocator = static_cast<c10::Allocator*>(PyLong_AsVoidPtr(allocator_ptr));
       PyDict_DelItemString(kwargs, "allocator");
+
+    } else if (device_ptr) {
+      // TODO: Need to change this to accept strings also
+      TORCH_CHECK(
+        THPDevice_Check(device_ptr), "invalid device");
+      PyDict_DelItemString(kwargs, "device");
+      c10::Device device = reinterpret_cast<THPDevice*>(device_ptr)->device;
+
+      if (device.type() == at::kCPU) {
+        allocator = c10::GetDefaultCPUAllocator();
+#ifdef USE_CUDA
+      } else if (device.type() == at::kCUDA) {
+        at::globalContext().lazyInitCUDA();
+        allocator = c10::cuda::CUDACachingAllocator::get();
+        device_guard.set_device(device);
+#endif
+      } else {
+        TORCH_CHECK(false,
+          THPStorageStr, "(): Storage device not recognized: ", device.type());
```

Comment:
TODO: make this work generically for all backends

Reply:
I initially tried using `c10::GetAllocator(device.type())`, but for some reason it returns the CPU allocator when `device.type() == at::kCUDA`. I could look into why that is and potentially fix it